doc/neps/nep-0042-new-dtypes.rst

```diff
@@ -0,0 +1,796 @@
+to be safe and NumPy will then ask ``ArbitraryWidthInteger`` to cast to the
+specifically requested instance.
+If ``ArbitraryWidthInteger`` has smaller representations, even ones that ``Int24``
+does not know exist,
+this provides the information that the cast is unsafe in the followup step.
+
+The main alternative to the proposed design is to move most of the information
+which is here pushed into the ``CastingImpl`` directly into methods
+on the DTypes. This, however, will not allow the close similarity between casting
+and universal functions. On the up side, it reduces the necessary indirection
+as noted below.
+
+An initial proposal defined two methods ``__can_cast_to__(self, other)``
+to dynamically return ``CastingImpl``.
+The advantage of this addition is that it removes the requirement to know all
+possible casts at DType creation time (of one of the involved DTypes).
+Such API could be added at a later time. It should be noted, however,
+that it would be mainly useful for inheritance like logic, which can be
+problematic. As an example two different ``Float64WithUnit`` implementations
+both could infer that they can unsafely cast between one another when in fact
+some conbinations should cast safely or preserve the Unit (both of which the
+"base" ``Float64`` would discard).
+In the proposed implementation this is not possible, since the two implementations
+are not aware of each other.
+
+
+**Notes:**
+
+The proposed ``CastingImpl`` this designed to be compatible with the
+``UFuncImpl`` proposed in NEP YY.
+While initially it will be a distinct object, the aim is that ``CastingImpl``
+can be a subclass of ``UFuncImpl``.
+Once this happens, this will naturally allow the use of a ``CastingImpl`` to
+pass around a specialized casting function directly if so wished.
+
+In the future, we may considering adding a way to spell out that specific
+casts are known to be *not* possible.
+
+
+C-Side API
+^^^^^^^^^^
+
+.. note:: At the time of writing, this API is a general design goal.
+          Due to the size of the proposed changes, details and names will be
+          in flux and updated as necessary.
+
+DType creation
+""""""""""""""
+
+As already mentioned in NEP 41, the interface to define new DataTypes in C
+is modelled after the limited API in Python, the above mentioned slots,
+and some additional necessary information will thus be passed within a slots
+struct and identified by ``ssize_t`` integers::
+
+    static struct PyArrayMethodDef slots[] = {
+        {NPY_dt_method, method_implementation},
+        ...,
+        {0, NULL}
+    }
+
+    typedef struct{
+      PyTypeObject *typeobj;    /* type of python scalar */
+      int flexible;             /* Is the dtype flexible? */
+      int abstract;             /* Is the dtype abstract? */
+      int flags;                /* Currently only the "needs API" flag */
+      /* NULL terminated CastingImpl; is copied and references are stolen */
+      CastingImpl *castingimpls[];
```

Comment:
You are not inheriting from anything except DType here (unless its abstract). So yes, I could make an `AbstractDType` as a subclass (or superclass) of `np.dtype`, which is used to indicate that it is abstract.
My current thinking is that I want to be able to ensure that any abstract dtype cannot be simply instantiated, and this might be easiest. But hmmmm. The `flexible` part is unnecessary, but should allow some optimizations and also sanity checks because "flexible" dtypes require more logic.

Reply:
We can use ABCs, I am not sure it is valuable though (plus I am a bit scared of python changing how they define ABCs). In fact ABCs are *effectively* defined in Python right now (they are optimized using C, but they do not use an `ExtenionMetaClass` as I do (that does not make things harder maybe though).

On the other hand, I expect the actual need we have is only overriding `isinstance`, etc. so I am not sure it is valuable to use ABCs, it seems like it may add more complexity then it solves.

Reply:
Indeed, not sure it adds enough, but really it was meant as encouragement (for you, I fear...) to look what they actually do so that we do not reinvent the wheel. The other part of ABCs is, of course, to ensure that any concrete instance has all the needed methods, etc., so they would ensure that abstract dtypes cannot be instantiated.

Reply:
Note to self: It could also be an option to include a version (maybe NPY_API_VERSION maybe another one) in these structs, that way we can add fields at to the end without breaking ABI (with the new fields being safe to ignore on old versions). Note sure I like it, and it would still only allow adding things at the end of the struct.